Source/DafnyCore/Verifier/BoogieGenerator.TrStatement.cs

```diff
@@ -1549,15 +1549,24 @@ public partial class BoogieGenerator {
           var toks = new List<IToken>();
           var types = new List<Type>();
           var decrs = new List<Expr>();
+          var oldDecreases = new List<Expression>();
           foreach (Expression e in theDecreases) {
             toks.Add(e.tok);
             types.Add(e.Type.NormalizeExpand());
+            // Note: this label doesn't exist in the program, and is useful
+            // only for explanatory purposes.
+            var olde = new OldExpr(e.tok, e, "LoopEntry") {
+              Type = e.Type
+            };
+            oldDecreases.Add(olde);
```

Comment:
Wrapping the decreases expression with old seems like it would work, but the only problem with that is that it does not work for local variables, i.e. `old()` is currently only taking the old heap for references, but not for stack variables.
Ideally, you should also traverse "e" and replace all modified local variables (like `k`) by new ghost identifiers (like `prev_k`), and store the mapping in the PODesc so that the proof obligation description could describe not only the assertion of the decreases, but also the additional ghost assignments that need to be done at the beginning of the loop (`ghost var prev_k := k;`)

Reply:
Good catch!